pkg/operator/ceph/object/rgw.go

```diff
@@ -91,55 +116,160 @@ func (c *clusterConfig) createOrUpdate(update bool) error {
+		if daemonsetNum > 0 {
+			for i, d := range daemons.Items {
+				// Delete any existing daemonset
+				if err := k8sutil.DeleteDaemonset(c.context.Clientset, c.store.Namespace, d.Name); err != nil {
+					return fmt.Errorf("error during deletion of daemonset %s resource: %+v", d.Name, err)
+				}
+
+				// Create a new deployment accordingly
+				daemonLetterID := k8sutil.IndexToName(i)
+				// Each rgw is id'ed by <store_name>-<letterID>
+				daemonName := fmt.Sprintf("%s-%s", c.store.Name, daemonLetterID)
+				// resource name is rook-ceph-rgw-<store_name>-<daemon_name>
+				resourceName := fmt.Sprintf("%s-%s-%s", AppName, c.store.Name, daemonLetterID)
+				rgwConfig := &rgwConfig{
+					ResourceName: resourceName,
+					DaemonID:     daemonName,
+				}
+
+				deployment, err := c.startDeployment(rgwConfig)
+				if err != nil {
+					return err
+				}
+
+				resourceControllerOwnerRef := &metav1.OwnerReference{
+					UID:        deployment.UID,
+					APIVersion: "v1",
+					Kind:       "deployment",
+					Name:       rgwConfig.ResourceName,
+				}
+
+				// Generate the keyring after starting the replication controller so that the keyring may use
+				// the controller as its owner reference; the keyring is deleted with the controller
+				err = c.generateKeyring(resourceControllerOwnerRef)
+				if err != nil {
+					return fmt.Errorf("failed to create rgw keyring. %+v", err)
+				}
+
+				// Generate the mime.types file after the rep. controller as well for the same reason as keyring
+				if err := c.generateMimeTypes(resourceControllerOwnerRef); err != nil {
+					return fmt.Errorf("failed to generate the rgw mime.types config. %+v", err)
+				}
+
+			}
 		}
-	}
-
-	// start the deployment or daemonset
-	var uid types.UID
-	var controllerType string
-	if c.store.Spec.Gateway.AllNodes {
-		daemonSet, err := c.startDaemonset()
+		// Remove the old rgw key once all the old rgws (daemonsets and deployments) have been replaced
+		err = client.AuthDelete(c.context, c.store.Namespace, oldRgwKeyName)
 		if err != nil {
 			return err
 		}
-		uid = daemonSet.UID
-		controllerType = "daemonset"
 	} else {
-		deployment, err := c.startDeployment()
-		if err != nil {
-			return err
+		// start a new deployment and scale up
+		desiredRgwInstances := int(c.store.Spec.Gateway.Instances)
+		for i := 0; i < desiredRgwInstances; i++ {
+			daemonLetterID := k8sutil.IndexToName(i)
+			// Each rgw is id'ed by <store_name>-<letterID>
+			daemonName := fmt.Sprintf("%s-%s", c.store.Name, daemonLetterID)
+			// resource name is rook-ceph-rgw-<store_name>-<daemon_name>
+			resourceName := fmt.Sprintf("%s-%s-%s", AppName, c.store.Name, daemonLetterID)
+
+			rgwConfig := &rgwConfig{
+				ResourceName: resourceName,
+				DaemonID:     daemonName,
+			}
+
+			// if intended to update, remove the old pods so they can be created with the new spec settings
+			if update {
+				err := k8sutil.DeleteDeployment(c.context.Clientset, c.store.Namespace, rgwConfig.ResourceName)
```

Comment:
That's not me, this code was already there, just a re-indent.

Reply:
Ok, how about if you remove it?

Reply:
If we remove it, new specs won't be propagated to existing pods, I guess we should wait for `UpdateDeploymentAndWait()` to be implemented so that can be removed safely. Agreed?

Reply:
An update already happens inside `startDeployment()` so I believe we're already covered. It will just get better when `UpdateDeploymentAndWait()` is implemented.

Reply:
Coming next :)